src/panels/lovelace/cards/hui-shopping-list-card.ts

```diff
@@ -250,6 +301,49 @@ class HuiShoppingListCard extends SubscribeMixin(LitElement)
     }
   }
 
+  private async _toggleReorder() {
+    if (!Sortable) {
+      const sortableImport = await import(
+        "sortablejs/modular/sortable.core.esm"
+      );
+      Sortable = sortableImport.Sortable;
+    }
+    this._reordering = !this._reordering;
+    await this.updateComplete;
+    if (this._reordering) {
+      this._createSortable();
+    } else {
+      this._sortable?.destroy();
+      this._sortable = undefined;
+    }
+  }
+
+  private _createSortable() {
+    this._sortable = new Sortable(this.shadowRoot!.getElementById("sortable"), {
+      animation: 150,
+      fallbackClass: "sortable-fallback",
+      dataIdAttr: "item-id",
+      handle: "ha-svg-icon",
+      onSort: async (evt) => {
+        reorderItems(this.hass!, this._sortable.toArray()).catch(() =>
+          this._fetchData()
+        );
+        // Move the shopping list item in memory.
+        this._uncheckedItems!.splice(
+          evt.newIndex,
+          0,
+          this._uncheckedItems!.splice(evt.oldIndex, 1)[0]
+        );
+        this._renderEmptySortable = true;
```

Comment:
I think we should also do a round of `_renderEmptySortable` when `this._uncheckedItems` is changed? What happenes when you add an item while you are sorting the list? And what happens when storing the new order (line 329) failed?

Reply:
Tested the 2 scenarios without changing anything, both works fine.

When `this._uncheckedItems` is changed, the whole `sortable` div content will be re-rendered (at [line 150](https://github.com/home-assistant/frontend/pull/7296/files#diff-3f9c900506d8e374961fa5be5db01f33e7d214a7e3f02ecd9e0dcb825fc3b993R150)), so everything should be straight.

Reply:
That is not true, Lit uses comments to keep track of things, because with sorting we change the DOM but not the comments Lit loses track of some elements. That's why we clear the DOM and render empty and start over. That can cause duplicate items in the list when not rendering empty.

Reply:
I'll do more digging tonight.

Reply:
@bramkragten I took another look, I still think there is no need to explicitly 'do a round of `_renderEmptySortable` when this._uncheckedItems is changed'.

Because at [line 150](https://github.com/home-assistant/frontend/pull/7296/files#diff-3f9c900506d8e374961fa5be5db01f33e7d214a7e3f02ecd9e0dcb825fc3b993R150): `guard([this._uncheckedItems, this._renderEmptySortable]...` decides that when either one of the 2 values (`_uncheckedItems`, `_renderEmptySortable`) changes, the content of sortable `div` will be re-rendered ([doc](https://lit-html.polymer-project.org/guide/template-reference#guard)). 

When add an item while sorting the list, `_uncheckedItems` is changed therefore the content of sortable `div` will be re-rendered. Unless you are talking about the scenario that: the user adds an item after the user starts dragging and BEFORE the user puts it down, but I don't see how that is possible.
Also when storing the new order failed, `_uncheckedItems` is changed therefore the content of sortable `div` will be re-rendered.

In short: the list is messed up only after drag-n-drop, we do a round of `_renderEmptySortable` to re-render the list (also clean up the zombie elements). In any other cases, the change of `this._uncheckedItems` will always re-render list, no need to use `_renderEmptySortable`, also there is no zombie elements generated by drag-n-drop.

Reply:
But while I was digging into it, I found another edge case:

1. Drag an item around and put the item back to its original position.
2. The Lit format is messed up by dragging around.
3. But the cleanup block (toggling `_renderEmptySortable)` won't be executed because putting the item back to its original position don't trigger `onSort` event.
4. Therefore the format won't be cleaned up and causing issues in other user interactions.

The bug:

![Kapture 2020-11-27 at 16 00 25](https://user-images.githubusercontent.com/7324708/100486498-b64d2180-30c9-11eb-9f80-c3c645d38a53.gif)

The fix I put in is that instead of observing `onSort` event, we observe `onEnd` event, which will be triggered even if an item is dragged and put back to its original position. In the block, we check on `evt.oldIndex !== evt.newIndex` to decide if a websocket message needs to be sent.

After fix:

![Kapture 2020-11-27 at 16 06 30](https://user-images.githubusercontent.com/7324708/100486674-881c1180-30ca-11eb-97ef-9e824d7af2df.gif)

Reply:
The problem is that re-rendering doesn't clear up the nodes that have no comment anymore.

Reply:
@bramkragten now I see what your concern is, and it's a very valid point.

My thinking is: the no-comment-nodes are only generated after drag&drop, so we only need to clear them up after drag&drop. We are already doing that in the sortable's `onEnd` event. That should guarantee we always clear up the no-comment-nodes.

If you still don't think it's safe, I have no problem adding the clear-up when `this._uncheckedItems` changes. Let me know.

BTW I need to work on the toast to show errors, I forgot it last time but now I remember.

Reply:
@bramkragten  Thoughts? Does above make sense to you? Or should I go ahead add the clear-up when `this._uncheckedItems` changes? I don't mind either way.